src/Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs

```diff
@@ -1411,6 +1411,206 @@ public IActionResult SplitCoins([FromBody] SplitCoinsRequest request)
             }
         }
 
+
+        /// <summary>Splits and distributes UTXOs across wallet addresses</summary>
+        [HttpPost]
+        [Route("distribute-utxos")]
+        public IActionResult DistributeUtxos([FromBody] DistributeUtxosRequest request)
+        {
+            Guard.NotNull(request, nameof(request));
+
+            // checks the request is valid
+            if (!this.ModelState.IsValid)
+            {
+                return ModelStateErrors.BuildErrorResponse(this.ModelState);
+            }
+
+            DistributeUtxoModel model = new DistributeUtxoModel()
+            {
+                WalletName = request.WalletName,
+                UseUniqueAddressPerUtxo = request.UseUniqueAddressPerUtxo,
+                UtxosCount = request.UtxosCount,
+                UtxoPerTransaction = request.UtxoPerTransaction,
+                TimestampDifferenceBetweenTransactions = request.TimestampDifferenceBetweenTransactions,
+                MinConfirmations = request.MinConfirmations,
+                DryRun = request.DryRun
+            };
+
+            try
+            {
+                var walletReference = new WalletAccountReference(request.WalletName, request.AccountName);
+
+                Wallet wallet = this.walletManager.GetWallet(request.WalletName);
+                HdAccount account = wallet.GetAccount(request.AccountName);
+
+                List<HdAddress> addresses;
+                if (request.ReuseAddresses)
+                {
+                    addresses = this.walletManager.GetUnusedAddresses(walletReference, request.UseUniqueAddressPerUtxo ? request.UtxosCount : 1, request.UseChangeAddresses).ToList();
+                }
+                else if (request.UseChangeAddresses)
+                {
+                    
+                    addresses = account.InternalAddresses.Take(request.UseUniqueAddressPerUtxo ? request.UtxosCount : 1).ToList();
```

Comment:
The list of Internal and External addresses could be smaller than `request.UtxosCount`.

Reply:
What if you have a UTXO per address?

Reply:
@maciejzaleski I recall I saw something like "GetUnusedAddress where you can specifiy how many address you want, and if they are not yet in the wallet it will create them, so you should use that

Reply:
@dangershony This code snippet gets a list of addresses for the coin distribution; I don't think I understand your question.
@MithrilMan  Yes, this is easy to fix, I've just made a note to address this edge case.

Reply:
Ah maybe I did not understand you comment 
>The list of Internal and External addresses could be smaller than  request.UtxosCount

What do you mean here?

Reply:
The lists returned by the `Internal` and `External` properties are limited in the number of entries. Those properties are not generating new addresses but rather return whatever addresses we already hold in the wallet.
In case the API request requires to distribute the wallet funds to excessively large number of internal/external addresses the method will fail as it will only be able to provide a mach smaller set of addresses.
The solution is to generate more addresses with the `GetUnusedAddress` as @MithrilMan is suggesting

Reply:
AH I see so basically you want more addresses to distribute to.
FYI It's not a requirement to have an address per UTXO, its more recommended for privacy but not mandatory (or essential).

I would perhaps suggest more control to the user and pass a parameter that will specify the number of addresses to distribute to.